Enhancement #6: As a QA engineer, I want the plugin versions to be automatically handled so that I can easily run tests on a selected version

Miraeld opened the pull request:
# Description

Fixes #6 
Eliminates manual plugin version management for E2E tests. QA engineers can now configure versions in `plugin.config.ts` and the framework automatically downloads/builds required files. Supports version numbers, GitHub branches/tags, and direct URLs. Tests auto-check and download missing plugins before running. Fully backward compatible, existing workflows continue to work unchanged.

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## Detailed scenario

### What was tested

All new `npm` commands

### How to test

1. Copy `config/plugin.config.sample.ts` to `config/plugin.config.ts`
2. Run `npm run plugin:setup` to download/build configured plugin versions
3. Verify with `npm run plugin:list` - should show 3 plugin files
4. Run `npm run test:e2e` to verify automatic plugin setup in tests
5. Test force rebuild: `npm run plugin:setup -- --force`
 
## Technical description

### Documentation

This pull request introduces automated management of WP Rocket plugin versions for E2E testing, eliminating the need for manual setup of plugin zip files. It adds a configuration-driven plugin manager that can automatically download, build, validate, and clean required plugin versions, along with CLI commands and documentation updates to support this workflow.

**Automated Plugin Version Management**

- Added a new plugin manager (`utils/plugin-manager.ts`) that automatically downloads or builds required WP Rocket plugin versions based on a configuration file, replacing the previous manual process. The manager supports fetching by version, GitHub branch/tag, or direct URL, and provides functions for setup, validation, listing, and cleaning plugin files.
- Introduced a CLI tool (`plugin-manager-cli.ts`) with commands to set up, list, clean, and validate plugin versions, integrated into npm scripts for easy use (`plugin:setup`, `plugin:list`, `plugin:clean`, `plugin:validate`). [[1]](diffhunk://#diff-bbbe3ed0d2233eb6623d86fb39ef52714bd8722648879e5f845d37d2e7afc867R1-R93) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R33-R42)

**Configuration and Documentation**

- Added a sample configuration file (`config/plugin.config.sample.ts`) for specifying which plugin versions to use, with support for version numbers, GitHub branches/tags, or URLs. Users copy and edit this file to control test versions.
- Updated `README.md` to document the new automated plugin management workflow, configuration steps, and available CLI commands, replacing the old manual instructions. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L7-R17) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R33-R36)

**Integration with Test Lifecycle**

- Updated test hooks (`src/support/hooks.ts`) to automatically validate and set up plugin versions before running tests, ensuring all required plugin files are present and up-to-date. [[1]](diffhunk://#diff-ed33bcf8419693abda7a86b9fc828c5e762fb6cd5101338898c04c91df14fb99R28) [[2]](diffhunk://#diff-ed33bcf8419693abda7a86b9fc828c5e762fb6cd5101338898c04c91df14fb99R69-R78)

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification

No test for this

Miraeld committed:
feat: add plugin version manager core

- Add plugin.config.sample.ts for version configuration
- Add plugin-manager.ts utility for automatic plugin download/build
- Add plugin-manager-cli.ts for manual plugin management
- Support version numbers, GitHub branches/tags, and direct URLs

Closes #6

Miraeld committed:
feat: integrate plugin manager into test hooks

- Auto-check and download/build missing plugins in BeforeAll hook
- Tests now automatically setup required plugin versions
- No manual intervention needed

Miraeld committed:
feat: add npm scripts for plugin management

- Add plugin:setup - Setup all configured versions
- Add plugin:list - List plugin files status
- Add plugin:clean - Remove all plugin files
- Add plugin:validate - Check required files exist
- Add open-report script for convenience

Miraeld committed:
chore: update .gitignore for plugin manager

- Ignore plugin.config.ts (user-specific configuration)
- Ignore .tmp-plugin-build (temporary build directory)

Miraeld committed:
docs: update README for automatic plugin management

- Remove manual plugin download/rename instructions
- Add automatic plugin management section
- Reference plugin.config.sample.ts for configuration examples
- List CLI commands for optional manual management

Miraeld committed:
fix: Add TypeScript return types and improve error handling in CLI

- Add explicit return types to main() and printHelp() functions
- Fix lexical declaration in case block by adding braces
- Improve error handling with proper TypeScript type checking

Miraeld committed:
fix: Add input validation and security improvements to plugin manager

Security enhancements:
- Add path traversal protection for downloadFile function
- Add version format validation for downloadFromReleases
- Add command injection prevention with sanitizeShellInput
- Fix GitHub token exposure (use git credential helper)
- Add redirect location validation

Code quality improvements:
- Fix race condition in error handler (use try-catch)
- Remove unused pluginDirName variable
- Fix force parameter usage (use parameter instead of global)
- Update CLI flag documentation (--force vs --force-plugin-rebuild)
- Add graceful handling for missing config file
- Improve error handling with proper TypeScript types

Miraeld committed:
docs: Update plugin config sample with beta version note

- Add note about pre-release versions and alternative approaches
- Change default newRelease to stable version (3.20.2)
- Remove unused forceRebuild export (now handled internally)
- Update CLI flag documentation to match actual implementation

Miraeld commented:
All feedback has been addressed:

**Security fixes in [7dedcb8](https://github.com/wp-media/wp-rocket-e2e/commit/7dedcb8):**
- ✅ Command injection vulnerability (version validation) - Added input validation for `downloadFromReleases`
- ✅ Path traversal vulnerability - Added destination path validation in `downloadFile`
- ✅ GitHub token exposure - Now using git credential helper instead of embedding token in URL
- ✅ Command injection in buildFromGitHub - Added `sanitizeShellInput` function with pattern validation
- ✅ Race condition in error handler - Using try-catch instead of existsSync check
- ✅ Missing redirect location validation - Added null check before using redirect URL
- ✅ Missing config file handling - Added graceful error with helpful message

**Code quality fixes:**
- ✅ Force parameter usage in [7dedcb8](https://github.com/wp-media/wp-rocket-e2e/commit/7dedcb8) - Now properly uses parameter via internal flag
- ✅ Inconsistent flag documentation in [7dedcb8](https://github.com/wp-media/wp-rocket-e2e/commit/7dedcb8) - Updated to `--force` throughout
- ✅ Unused variable in [7dedcb8](https://github.com/wp-media/wp-rocket-e2e/commit/7dedcb8) - Removed `pluginDirName`
- ✅ Error handling in [69e5ee8](https://github.com/wp-media/wp-rocket-e2e/commit/69e5ee8) and [7dedcb8](https://github.com/wp-media/wp-rocket-e2e/commit/7dedcb8) - Using proper TypeScript type checking
- ✅ Beta version note in [68151aa](https://github.com/wp-media/wp-rocket-e2e/commit/68151aa) - Added documentation about pre-release versions

All ESLint errors are now resolved as well.

jeawhanlee reviewed (commented):
Great work on this Sir 🙌
Would it also be possible to enhance this to accept arguments of plugin versions directly from the CLI that way we might not need to modify the config file all the time for each test run or maybe this doesn't make sense.

CC @wp-media/qa-team Do you see any usefulness in this suggestion?

jeawhanlee commented:
To make things a bit more automated, Would it make sense to have an endpoint we can connect to that would always return the latest version and the previous stable version of WPR, that way, we'll eliminate configs

CC @wp-media/qa-team @wordpressfan

jeawhanlee reviewed (approved):
Approving this PR with need for some enhancement mentioned [here](https://github.com/wp-media/wp-rocket-e2e/issues/315)

Miraeld committed:
feat: add CLI version override system for plugin versions

- Create version-override module with support for multiple input formats
- Support CLI arguments (--previous-stable, --new-release, --specific-version)
- Support environment variables (PREVIOUS_STABLE, NEW_RELEASE, SPECIFIC_VERSION)
- Support npm config style arguments
- Add version format validation and display helpers

Miraeld committed:
feat: integrate CLI overrides into plugin manager

- Update plugin manager to accept version overrides
- Apply runtime overrides to base configuration
- Display active overrides when detected
- Add version labels for better logging
- Auto-trigger download/build for overridden versions

Miraeld committed:
feat: enable CLI overrides in test hooks

- Parse version overrides before test execution
- Pass overrides to plugin setup process
- Seamlessly integrate with existing validation flow

Miraeld committed:
docs: add comprehensive CLI version override documentation

- Document all CLI argument formats and options
- Add examples for common testing scenarios
- Include environment variable usage
- Document supported version formats
- Provide real-world usage examples

Miraeld committed:
Merge branch 'develop' into enhancement/6-selectable-plugin-version

Miraeld commented:
## CLI Version Override 
I've implemented the feedback from @jeawhanlee about accepting plugin version arguments directly from the CLI without needing to modify the config file each time.

### What's New

**CLI Override Support** - You can now override plugin versions at runtime using multiple methods:

1. **NPM config arguments:**
   ```bash
   npm run test:e2e --wpr=3.16.4.1 --previous=branch:release/3.16.4
   ```

2. **Environment variables:**
   ```bash
   WPR_VERSION=tag:3.16.4.1 PREVIOUS_VERSION=3.16.3 npm run test:e2e
   ```

3. **Command line arguments:**
   ```bash
   npm run test:e2e -- --wpr=https://example.com/custom-build.zip
   ```

### Supported Formats

All the same formats from `plugin.config.ts` are supported:
- Version numbers: `3.16.4.1`
- GitHub branches: `branch:develop`, `branch:release/3.16.4`
- GitHub tags: `tag:3.16.4.1`
- Direct URLs: `https://example.com/plugin.zip`

### Implementation Details

**New Files:**
- `utils/version-override.ts` - Handles parsing and applying CLI overrides

**Modified Files:**
- `utils/plugin-manager.ts` - Integrated override system with `getActivePluginConfig()` and updated `setupPluginVersions()` to accept overrides
- `src/support/hooks.ts` - Added `parseVersionOverrides()` call in `BeforeAll` hook to apply CLI arguments before plugin setup
- `README.md` - Added comprehensive documentation with examples and usage scenarios

### Example Use Cases

**Test a specific version without config changes:**
```bash
npm run test:e2e --wpr=3.16.4.1
```

**Test a branch against previous stable:**
```bash
npm run test:e2e --wpr=branch:feature/new-optimization
```

**Override both versions for comparison testing:**
```bash
npm run test:e2e --wpr=branch:develop --previous=3.16.3
```

The config file still serves as the default, but CLI arguments take precedence when provided. This makes it easy to run quick tests without editing configuration files!

Mai-Saad commented:
- [ ] Run test with versions set in config files (no zips)
- [ ] Run test with versions set in config file and have already zips in plugin directory
- [ ] Run test with latest set in CLI and zips in plugin directory
- [ ] Run test with both latest and previous set in CLI
- [ ] Run test with both latest and previous set in CLI and report send to shared folder

Miraeld committed:
feat: implement priority-based plugin version management

- Add three-tier priority system for plugin version selection:
  1. CLI overrides (highest priority) - always clean and re-download
  2. Config file (medium priority) - clean and re-download from config
  3. Existing plugins (lowest priority) - use files already in folder

- Make config file optional - no longer fails if missing
- Add hasPluginConfigFile() to check config existence
- Update hooks.ts to handle all three cases with clear logging
- Validate plugin files and provide helpful error messages

Benefits:
- QA can use existing plugin files without re-downloading
- Developers can quickly override versions via CLI
- Config-based workflow still works as expected
- Clear priority system prevents confusion

Miraeld commented:
## Priority-Based Plugin Version Management

I've implemented a more flexible and QA-friendly approach to plugin version management based on team feedback. The system now uses a **three-tier priority system** that makes it easier to work with different testing scenarios.

### How It Works

The plugin manager now follows a clear priority order:

**1️⃣ CLI Overrides (Highest Priority)**
When you specify versions via command line, they take precedence over everything else:
- ✅ Plugin folder is **cleaned**
- ✅ Plugins are **re-downloaded** with your specified versions
- Overrides any config file settings

```bash
# Quick test with specific versions
npm run test:e2e -- --new-release=3.16.4 --previous-stable=3.16.3

# Test a branch
npm run test:e2e -- --new-release=branch:develop

# Mix different formats
npm run test:smoke -- --new-release=tag:3.16.4-beta5 --previous-stable=3.16.3
```

**2️⃣ Config File (Medium Priority)**
When NO CLI overrides are provided, but `config/plugin.config.ts` exists:
- ✅ Plugin folder is **cleaned**
- ✅ Plugins are **re-downloaded** from config
- Ensures fresh downloads every time

```bash
# Uses versions from config/plugin.config.ts
npm run test:e2e
```

**3️⃣ Existing Plugins (Lowest Priority - NEW!)**
When NO CLI overrides AND NO config file exist:
- ❌ Plugin folder is **NOT cleaned**
- ✅ Uses **existing plugin files** in `plugin/` folder
- Just validates that required files exist

```bash
# Manually place plugin files in plugin/ folder
ls plugin/
# previous_stable.zip  new_release.zip  wp-rocket_3.10.9.zip

# Run tests without re-downloading
npm run test:e2e
```

### Why This Is Better

**For QA:**
- ✅ Can use existing plugin files when testing (no re-download needed)
- ✅ Quick version switching via CLI without editing config
- ✅ Clear error messages when files are missing

**For CI/CD:**
- ✅ Predictable behavior with priority system
- ✅ Always get fresh downloads when using config
- ✅ Can use pre-downloaded plugins for faster runs

### Changes Made

**Modified files:**
- `src/support/hooks.ts` - Implements priority-based logic in BeforeAll hook
- `utils/plugin-manager.ts` - Made config optional, added `hasPluginConfigFile()`

**Commit:** [e429aa6](https://github.com/wp-media/wp-rocket-e2e/commit/e429aa6)

### Testing Examples

```bash
# Scenario 1: Use existing files (no config)
rm config/plugin.config.ts  # Remove config
npm run test:e2e            # Uses existing plugin files

# Scenario 2: Use config file
cp config/plugin.config.sample.ts config/plugin.config.ts
npm run test:e2e            # Cleans and re-downloads from config

# Scenario 3: Override via CLI
npm run test:e2e -- --new-release=branch:feature/new-optimization
# Cleans and downloads the branch version
```

Miraeld committed:
refactor: remove unused specificVersion configuration

- Remove specificVersion field from PluginVersionConfig interface
- Remove specificVersion from VersionOverrides interface
- Remove specificVersion file mapping from PLUGIN_FILES
- Remove specificVersion processing in setupPluginVersions
- Update all examples in sample config file
- Cleanup validatePluginFiles, listPluginFiles and cleanPluginFiles

Miraeld commented:
## Update: Removed `specificVersion` Configuration

**Commit:** [87ace3a](https://github.com/wp-media/wp-rocket-e2e/pull/312/commits/87ace3ad169e153faac909d69f5bb04874568ab2)

The `specificVersion` field has been removed from the plugin configuration as it was added by mistake and is not used by any test scenarios.

### Changes:
- ❌ Removed `specificVersion` from `PluginVersionConfig` interface
- ❌ Removed `specificVersion` from `VersionOverrides` interface  
- ❌ Removed `specificVersion` file mapping from `PLUGIN_FILES`
- ❌ Removed `specificVersion` processing from all plugin manager functions
- ✅ Updated sample configuration file to reflect simplified structure

### Why?
After investigation, no test features actually require the specific version file (`wp-rocket_3.10.9.zip`):
- The roll-back test doesn't use it
- The upgrading-plugin tests only use `previous_stable` and `new_release`
- The field was optional and appears to have been legacy/unused configuration

The plugin manager now only handles the two versions that are actually used in tests: `previousStable` and `newRelease`.

Miraeld committed:
feat: download plugin versions from GitHub releases

- Change download source from wp-rocket.me to GitHub releases archive
- Use GitHub's automatic source archive generation (archive/refs/tags/v{version}.zip)
- Requires GITHUB_TOKEN for private repository access
- Update documentation to reflect GitHub as download source
- Add 'v' prefix to version tags when fetching from GitHub

This ensures plugin downloads remain private to the organization
while allowing automated version management for E2E tests.

Miraeld commented:
## Update: Plugin Downloads Now Use GitHub Releases

**Commit:** 7b66bc4

### Changes

The plugin version management now downloads plugin versions from **GitHub releases** instead of wp-rocket.me, keeping downloads private to the organization.

#### What Changed:

1. **Download Source**: Version numbers (e.g., `3.20.1`) now download from `https://github.com/wp-media/wp-rocket/archive/refs/tags/v{version}.zip`
2. **Authentication**: Requires `GITHUB_TOKEN` to access private repository releases
3. **Automatic Archives**: Uses GitHub's automatically generated source archives (always available for any tag)
4. **Tag Format**: Automatically adds `v` prefix to version numbers (e.g., `3.20.1` → `v3.20.1`)

#### Configuration Required:

```typescript
export const pluginConfig: PluginVersionConfig = {
    previousStable: '3.20.1',
    newRelease: '3.20.2',
    repository: {
        owner: 'wp-media',
        name: 'wp-rocket',
        token: process.env.GITHUB_TOKEN  // Required for downloads
    }
};
```

#### Why This Change?

- ✅ **Security**: Downloads remain private within wp-media organization
- ✅ **Reliability**: GitHub source archives are automatically generated for every tag
- ✅ **Consistency**: Uses same authentication mechanism as branch/tag builds
- ✅ **No Public Access**: Prevents unauthorized downloads of WP Rocket

#### Supported Version Formats:

- `3.20.1` - Downloads from GitHub releases (requires GITHUB_TOKEN)
- `branch:develop` - Clones and builds from branch (requires GITHUB_TOKEN)
- `tag:v3.20.1` - Clones and builds from tag (requires GITHUB_TOKEN)
- `https://example.com/file.zip` - Direct URL download

All automated methods now require proper GitHub authentication, ensuring plugin files remain secure.

Miraeld committed:
feat: extract and build GitHub source archives with dependencies

- Downloads are now extracted to temp directory
- Runs composer install --no-dev --no-scripts to install dependencies
- Skips post-install scripts that require dev dependencies (Mozart)
- Manually runs composer dump-autoload for optimized autoloading
- Checks for npm dependencies and runs build scripts if present
- Creates final plugin zip with all built dependencies included

This ensures version downloads have vendor/ directory and built assets,
making them ready for use in E2E tests without missing dependencies.

Miraeld commented:
## Update: Automatic Build Process for Version Downloads

**Commit:** c03ebad

### Issue

GitHub's source archives contain only raw source code without built dependencies:
- No `vendor/` directory (Composer dependencies)
- No `node_modules/` (npm dependencies)
- No built assets (CSS/JS)

### Solution

Version downloads now automatically build the plugin after downloading:

1. **Download** source archive from GitHub releases
2. **Extract** to temporary directory
3. **Install Composer dependencies**: `composer install --no-dev --no-scripts`
   - Skips post-install scripts that require dev dependencies (like Mozart)
   - Manually runs `composer dump-autoload` for optimized autoloading
4. **Install npm dependencies**: `npm install` (if package.json exists)
5. **Build assets**: Runs build scripts if available (e.g., `build:css`, `build:js`)
6. **Create plugin zip** with all built dependencies included

### Result

- Downloaded plugins now include `vendor/` directory (~600KB of dependencies)
- Plugins are ready for use in E2E tests without missing dependencies
- No manual build steps required

Miraeld committed:
style: fix ESLint naming-convention errors for HTTP headers

Add inline eslint-disable comments for HTTP header property names
that use standard capitalization (User-Agent, Authorization).

jeawhanlee commented:
To be honest, I don't know how I feel about compiling the plugin on e2e, we might need to brainstorm on this.

Miraeld commented:
@jeawhanlee I kinda agree, but we can't download it from wp-rocket.me as it's subject to authentification.
